### PR TITLE
feat: String interpolation

### PR DESCRIPTION
**Pull request by developerfred**
base https://github.com/AssemblyScript/assemblyscript/pull/1115
close https://github.com/near/near-sdk-as/issues/201

- [x] merge
- [ ] update test

### CODE REVIEW

File: tests/compiler/std/string.ts [hunk 1]
```diff
@@ -731,10 +733,118 @@ assert(dtoa(0.000035689) == "0.000035689");
...
+  if (origUpperCode != expectUpperCode) {
+    trace("origUpperCode != expectUpperCode", 3, i, <f64>origUpperCode, <f64>expectUpperCode);
+ }
+
+  assert(origLowerCode == expectLowerCode);
+  assert(origUpperCode == expectUpperCode);
+}
+
+
+
```
Comment (MaxGraey) [inline tests/compiler/std/string.ts:842]: All this block of tests transferred to `tests/compiler/std/string-casemapping.ts`. I guess it just appear after conflict merging

Comment (developerfred) [reply tests/compiler/std/string.ts:842]: I will check now, thanks @MaxGraey

**Comment by pulpdrew** [comment]
I was also working on this, before realizing that someone else was. I'll leave it to you, but here are some test cases I came up while testing my implementation, just in case they're helpful:

tests/parser/literals.ts:
```
``;
`123`;
`$1`;
`$ {}`;
`} {`;
`${foo}`;
`${foo}${foo}`;
`1${foo() + 1}`;
`${2 - 1}2`;
`12${3}45`;
`\${} ${foo}`;
`1${`2${3}`}`;
`foo:

${foo}`;
`${}` // Error
```

tests/parser/literals.fixture.ts (these will probably differ for your implementation)
```
"123";
"1\"23";
"1\"2\\3";
"\0\n\\n\r";
"";
"123";
"$1";
"$ {}";
"} {";
(foo).toString();
(foo).toString() + (foo).toString();
"1" + (foo() + 1).toString();
(2 - 1).toString() + "2";
"12" + (3).toString() + "45";
"${} " + (foo).toString();
"1" + ("2" + (3).toString()).toString();
"foo:\n\n" + (foo).toString();
"" // Should write out an error, too
```

**Comment by MaxGraey** [comment]
Btw will be great if you avoid using `String#concat` and use [joinStringArray](https://github.com/AssemblyScript/assemblyscript/blob/master/std/assembly/util/string.ts#L972) instead [here](https://github.com/AssemblyScript/assemblyscript/pull/1427/files#diff-6518fffc4e430fcc63cd30eb07131a95R3892)

**Comment by developerfred** [comment]
> Btw will be great if you avoid using `String#concat` [here](https://github.com/AssemblyScript/assemblyscript/pull/1427/files#diff-6518fffc4e430fcc63cd30eb07131a95R3892) and use [joinStringArray](https://github.com/AssemblyScript/assemblyscript/blob/master/std/assembly/util/string.ts#L972) instead when count of arguments more than 2

thanks, my return parts look like this 
```
return parts.reduce((acc: Expression | null, expr: Expression) => {
      if (acc == null) return expr;
      return Node.createTemplateLiteralExpression(joinStringArray(expr), tn.range(startPos, tn.pos));
    }, null);
```

**Comment by MaxGraey** [comment]
I guess it should be in pseudocode:
```ts
if (parts.lenght == 0) return <empty string>;
else if (parts.lenght == 1) {
  return isString(parts[0]) 
       ? parts[0] 
       : makeCall('toString', parts[0])
} else if (parts.lenght  == 2) {
  return Node.createBinaryExpression(Token.PLUS, parts[0], parts[1], tn.range(startPos, tn.pos));
} else {
  return makeCall('joinStringArray', makeStringArray(parts));
}
```

**Comment by developerfred** [comment]
@MaxGraey  According to the pseudocode
- [x] `isStringExpression(expr: any): boolean `

- [x] `makeCallExpression(callExpression: string , ...args:any)`



```javascript
    // isStringExpression checks if the string is an Expression
    function isStringExpression(expr: any): boolean {
      return (expr instanceof Expression) ? true : false
    };

```

    
```javascript
    // makeCallExpression calls an expression according to the function
    function makeCallExpression(callExpression: string , ...args:any) {
      switch (callExpression) {
        case 'toString':
          return (typeof args == 'object') ? JSON.stringify(args) : args.toString()
          break;
        case 'joinThreeStrings':
          return joinThreeStrings(args[0], args[1], args[2]);
          break;    
        case 'joinStringArray':
          return joinStringArray(args);
          break;          
      }
```

**Comment by MaxGraey** [comment]
No, this will much more complicated. I guess @dcodeIO will tell you better how to implement such routines

**Comment by dcodeIO** [comment]
To emit a call to a standard library function, one would obtain a reference to it from the program, resolve it and create a direct call:

```ts
let prototype = program.require("~lib/file/concat3", ElementKind.FUNCTION_PROTOTYPE);
let instance = assert(resolver.resolveFunction(<FunctionPrototype>prototype, null)); // here: not generic
let expr = compiler.makeCallDirect(instance, [arg1, arg2, arg3], reportNode, immediatelyDropped?, skipAutorelease?);
```

emitting about

```wat
(call $~lib/file/concat3
 (arg1)
 (arg2)
 (arg3)
)
```

**Comment by developerfred** [comment]
@dcodeIO Thank you so much!

**Comment by MaxGraey** [comment]
Already implemented in #1715. Closing

**Pull request closed without merging**